### PR TITLE
Moved iEnvoyProxy services into Envoy itself

### PR DESCRIPTION
**Pull request by mnbogner**
Moved iEnvoyProxy services down into Envoy itself to streamline integration.  Now only a list of urls (with certain formatting requirements) and an optional url to test direct connections are required as arguments and any required service will be started or stopped by Envoy itself as the urls are verified.

Additional parameters may be required when building to use the new features.  Specifically, Hysteria requires a key and a certificate and dnstt requires several configuration values.  These parameters can be passed in on the command line or added to the project settings in Android Studio.

I will draft an updated readme before merging this branch.

### CODE REVIEW

**Comment by mnbogner** [comment]
There is some current behavior I wanted to highlight.  I don't know that it's right or wrong but it might be worth discussing whether it might cause unwanted friction for users.
 - As urls are found to be valid or invalid, they are added to a corresponding list and then that entire list is returned.  This means that the urls should be ordered roughly according to speed.  There is a hard-coded delay for urls that require services to allow those services to start.  There is also a smaller hard coded delay for http(s) urls to allow a chance to test a direct connection.  The net effect of this is that it is up to the app itself to decide what to do when each list is received.
 - If all urls are invalid (or if no urls were submitted) Envoy will try to fetch additional urls with dnstt.  Currently there is no broadcast to indicate this behavior, and the number and details of the urls returned may be unknown to the app. (ie: if you submit 6 urls, you may receive a list of 6 invalid urls, then receive a new list of 1 or more valid urls)  This may lead to unpredictable behavior.
 - If any parameters required to start a service are missing, the corresponding urls will not be tested and there will be no status broadcast.

**Review by stevenmcdonald — APPROVED** [review]
With the caveat that this isn't exactly my area of expertise, everything looks good to me

**Comment by stevenmcdonald** [comment]
I do get some warnings from the build:

> w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/CronetUrlRequestCallback.kt: (151, 30): 'SPDY_3' is deprecated. OkHttp has dropped support for SPDY. Prefer {@link #HTTP_2}.
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (155, 32): Parameter 'index' is never used, could be renamed to _
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (263, 13): Condition 'v2wsParts == null' is always 'false'
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (286, 13): Condition 'v2srtpParts == null' is always 'false'
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (309, 13): Condition 'v2wechatParts == null' is always 'false'
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (491, 56): Unnecessary non-null assertion (!!) on a non-null receiver of type JSONArray!
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (641, 25): Unnecessary safe call on a non-null receiver of type UrlResponseInfo?
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (643, 71): Unnecessary safe call on a non-null receiver of type UrlResponseInfo?
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (643, 96): Unnecessary safe call on a non-null receiver of type UrlResponseInfo?
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (661, 71): Unnecessary safe call on a non-null receiver of type UrlResponseInfo?
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (661, 96): Unnecessary safe call on a non-null receiver of type UrlResponseInfo?
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (679, 30): Parameter 'request' is never used
w: /build/envoy/android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt: (679, 52): Parameter 'info' is never used

Some of that stuff might be worth looking in to

**Comment by stevenmcdonald** [comment]
oh, I just notices this too:

> > Configure project :envoy
WARNING: no dnsttserver prop found
WARNING: no dnsttkey prop found
WARNING: no dnsttpath prop found
WARNING: no dohUrl prop found
WARNING: no dotAddr prop found
WARNING: no hystcert prop found

How do those get passed in for command line builds, can you just pass the `-P ...` params to the `gradlew`? We should update `android/build-envoy.sh` to pass them in... I guess from ENV vars?

**Review by stevenmcdonald — CHANGES_REQUESTED** [review]
I think we need to update the build scripts before merge

**Comment by mnbogner** [comment]
> I think we need to update the build scripts before merge

yes

File: android/README.md [hunk 1]
```diff
@@ -16,73 +25,79 @@ Build the demo module to see it in action, or just call this in `Activity`'s `on
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        LocalBroadcastManager.getInstance(this).registerReceiver(mBroadcastReceiver, IntentFilter().apply {
+            addAction(org.greatfire.envoy.BROADCAST_URL_VALIDATION_SUCCEEDED)
+            addAction(org.greatfire.envoy.BROADCAST_URL_VALIDATION_FAILED)
+        })
+    
+        val listOfUrls = mutableListOf<String>()
+        listOfUrls.add(foo)
```
Comment (vitriolix) [inline android/README.md:111]: um...

File: android/README.md [hunk 2]
```diff
@@ -16,73 +25,79 @@ Build the demo module to see it in action, or just call this in `Activity`'s `on
...
+    
+ - submit(context: Context, urls: List<String>)
+ - submit(context: Context, urls: List<String>, directUrl: String)
+    
+If the optional directUrl parameter is included, Envoy will attempt to connect to that url directly first.  This can be included to avoid using proxy resources when the target domain is not blocked.
+
+## Basic envoy integration
+
+```kotlin
+    private val mBroadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
```
Comment (vitriolix) [inline android/README.md:75]: why the mB style capitalization?

Comment (mnbogner) [reply android/README.md:75]: it was part of the original example.  i can change it if you think it's important.

File: android/README.md (hunk 1 above)
Comment (mnbogner) [reply android/README.md:114]: i'll do some edits, but this is just an example so i'm not going to use actual urls or config parameters.

**Pull request merged**